pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,260 @@
+	crdList := &apiextv1.CustomResourceDefinitionList{}
+	if err := k8sClient.List(context.TODO(), crdList); err != nil {
+		return nil, fmt.Errorf("failed to get CRD list: %v", err)
+	}
+	for _, crd := range crdList.Items {
+		if strings.HasSuffix(crd.Name, "crd.antrea.io") {
+			crdNames.Insert(crd.Name)
+		}
+	}
+
+	return crdNames, nil
+}
+
+// getCRDNamesToUpgrade gets names of Antrea CRDs to upgrade.
+func getCRDNamesToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNamesToUpgrade sets.Set[string]) (sets.Set[string], error) {
+	antreaCRDNames, err := getAntreaCRDNames(k8sClient)
+	if err != nil {
+		return nil, err
+	}
+	// If the user-provided name list of CRDs to upgrade is empty, upgrade all Antrea CRDs.
+	if crdNamesToUpgrade.Len() == 0 {
+		crdNamesToUpgrade = antreaCRDNames
+	} else {
+		// If the user-provided name list of CRDs to upgrade is not empty, and it contains CRDs without suffix
+		// "crd.antrea.io", skip these CRDs.
+		for name := range crdNamesToUpgrade.Difference(antreaCRDNames) {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip CRD %q which is not created by Antrea.\n", name)
+		}
+		// Only upgrade the CRDs with suffix "crd.antrea.io".
+		crdNamesToUpgrade = crdNamesToUpgrade.Intersection(antreaCRDNames)
+	}
+	return crdNamesToUpgrade, nil
+}
+
+// getCRDsToUpgrade gets a list of Antrea CRDs to upgrade.
+func getCRDsToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) < 2 {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(cmd *cobra.Command, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := k8sClient.List(context.TODO(), list); err != nil {
+		return err
+	}
+	if len(list.Items) == 0 {
+		return nil
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Upgrading %d objects of CRD %q.\n", len(list.Items), crd.Name)
+	for _, obj := range list.Items {
+		copyObj := obj
+		err := retry.OnError(wait.Backoff{
+			Duration: time.Second,
+			Steps:    5,
+		}, func(err error) bool {
+			return validateErr(err) != nil
+		}, func() error {
+			return k8sClient.Update(context.TODO(), &copyObj)
```

Comment:
shouldn't it get the object once when retrying? otherwise this can never succeed since it has a stale resource version. Even if it succeeds by removing the resource version in the request, it would override the updates made after you list the objects.

Reply:
Updated here, every time when to update the object, get the object first by the namespace+name from the list.